pkg/ee/default-policy-controller/controller.go

```diff
@@ -0,0 +1,399 @@
+	reconciler := &Reconciler{
+		Client: mgr.GetClient(),
+
+		workerName:   workerName,
+		recorder:     mgr.GetEventRecorderFor(ControllerName),
+		seedGetter:   seedGetter,
+		configGetter: kubermaticConfigurationGetter,
+		log:          log,
+		versions:     versions,
+	}
+
+	_, err := builder.ControllerManagedBy(mgr).
+		Named(ControllerName).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: numWorkers,
+		}).
+		// Watch for clusters
+		For(&kubermaticv1.Cluster{}).
+		// Watch changes for PolicyTemplates that have been enforced.
+		Watches(&kubermaticv1.PolicyTemplate{}, enqueueClusters(reconciler, log), builder.WithPredicates(withEventFilter())).
+		Build(reconciler)
+
+	return err
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	cluster := &kubermaticv1.Cluster{}
+	if err := r.Get(ctx, request.NamespacedName, cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	if cluster.DeletionTimestamp != nil {
+		// Cluster deletion in progress, skipping reconciliation
+		r.log.Debugw("Cluster deletion in progress, skipping reconciliation", "cluster", cluster.Name)
+		return reconcile.Result{}, nil
+	}
+
+	result, err := util.ClusterReconcileWrapper(
+		ctx,
+		r,
+		r.workerName,
+		cluster,
+		r.versions,
+		kubermaticv1.ClusterConditionDefaultPolicyControllerReconcilingSuccess,
+		func() (*reconcile.Result, error) {
+			return r.reconcile(ctx, cluster)
+		},
+	)
+
+	if result == nil || err != nil {
+		result = &reconcile.Result{}
+	}
+
+	if err != nil {
+		r.recorder.Event(cluster, corev1.EventTypeWarning, "ReconcilingError", err.Error())
+	}
+
+	return *result, err
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, cluster *kubermaticv1.Cluster) (*reconcile.Result, error) {
+	// Ensure that the cluster is healthy first, this is important for the default policy controller, and for all policy controllers.
+	if !cluster.Status.ExtendedHealth.AllHealthy() {
+		r.log.Debugw("Cluster is not healthy, skipping reconciliation", "cluster", cluster.Name)
+		return &reconcile.Result{RequeueAfter: 10 * time.Second}, nil
+	}
+
+	// Determine whether to ignore default policies
+	//nolint:staticcheck
+	ignoreDefaultPolicies := false
+
+	// Default policies are already created.
+	if cluster.Status.HasConditionValue(kubermaticv1.ClusterConditionDefaultPolicyBindingsControllerCreatedSuccessfully, corev1.ConditionTrue) {
+		ignoreDefaultPolicies = true
+	}
+
+	// List all PolicyTemplates
+	policyTemplates := &kubermaticv1.PolicyTemplateList{}
+	if err := r.List(ctx, policyTemplates); err != nil {
+		return nil, fmt.Errorf("failed to list PolicyTemplates: %w", err)
+	}
+
+	if cluster.Status.NamespaceName == "" {
+		return nil, fmt.Errorf("cluster %s has no namespace name", cluster.Name)
+	}
```

Comment:
This is IMHO overly cautious. A cluster without namespace can never be AllHealthy. If the Cluster object is in such a bad shape that it pretend to be healthy but has no namespace, this controller is allowed to explode loudly.